pkg/reconciler/cache/publishedresources/replication/replication_controller.go

```diff
@@ -0,0 +1,224 @@
+		replicated:         replicated,
+		gvr:                gvr,
+		callback:           callback,
+	}
+
+	_, _ = c.replicated.Local.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueObject(obj, gvr) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueObject(obj, gvr) },
+		DeleteFunc: func(obj interface{}) { c.enqueueObject(obj, gvr) },
+	})
+
+	_, _ = c.replicated.Global.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueCacheObject(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueCacheObject(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueCacheObject(obj) },
+	})
+
+	return c, nil
+}
+
+func (c *Controller) enqueueAllCacheObjects() {
+	objs := c.replicated.Global.GetIndexer().List()
+	for _, obj := range objs {
+		c.enqueueCacheObject(obj)
+	}
+}
+
+// Idea is that locally we store object with the original GVR.
+// In the cache we store object with the CachedResource GVR, and store the original GVR in the labels.
+// This way we can extract the original GVR from the CachedResource and use it to enqueue the object.
+// Name is the name of the original object so by knowing gvr and name we can enqueue or get the object from the informers.
+
+func (c *Controller) enqueueObject(obj interface{}, gvr schema.GroupVersionResource) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	gvrKey := fmt.Sprintf("%s.%s.%s::%s", gvr.Version, gvr.Resource, gvr.Group, key)
+	c.queue.Add(gvrKey)
+}
+
+func (c *Controller) enqueueCacheObject(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	// This way we extract what is the original GVR of the object that we are replicating.
+	cr, ok := obj.(*cachev1alpha1.CachedResource)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expected *cachev1alpha1.CachedResource, got %T", obj))
+		return
+	}
+
+	labels := cr.GetLabels()
+	gvr := schema.GroupVersionResource{
+		Group:    labels[LabelKeyObjectGroup],
+		Version:  labels[LabelKeyObjectVersion],
+		Resource: labels[LabelKeyObjectResource],
+	}
+
+	gvrKey := fmt.Sprintf("%s.%s.%s::%s", gvr.Version, gvr.Resource, gvr.Group, key)
+	c.queue.Add(gvrKey)
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *Controller) Start(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(cacheclient.WithShardInContext(ctx, shard.New(c.shardName)), logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func (c *Controller) Stop() {
+	c.queue.ShutDown()
+}
```

Comment:
As mentioned in another comment this would let the controller run - maybe `.Start` should make a cancellable context and `.Stop` calls the cancel?

That would also kill the queue as the queue shutdown is deferred

Reply:
I think I gonna just kill `Stop` method all together and add cancel into `register` and kill it on `unregister`

Reply:
I like that; and it's a pattern that already exists e.g. in other controllers maintaining other controllers